include/mbedtls/gcm.h

```diff
@@ -253,22 +253,42 @@ int mbedtls_gcm_starts( mbedtls_gcm_context *ctx,
  *                  input buffer. If the buffers overlap, the output buffer
  *                  must trail at least 8 Bytes behind the input buffer.
  *
- * \param ctx       The GCM context. This must be initialized.
- * \param length    The length of the input data.
- * \param input     The buffer holding the input data. If \p length is greater
- *                  than zero, this must be a readable buffer of at least that
- *                  size in Bytes.
- * \param output    The buffer for holding the output data. If \p length is
- *                  greater than zero, this must be a writable buffer of at
- *                  least that size in Bytes.
+ * \param ctx           The GCM context. This must be initialized.
+ * \param input         The buffer holding the input data. If \p input_length
+ *                      is greater than zero, this must be a readable buffer
+ *                      of at least \p input_length bytes.
+ * \param input_length  The length of the input data in bytes.
+ * \param output        The buffer for the output data. If \p output_length
+ *                      is greater than zero, this must be a writable buffer of
+ *                      of at least \p output_size bytes.
+ *                      This function may withhold the end of the output if
+ *                      it is a partial block for the underlying block cipher.
+ *                      That is, if the cumulated input passed to
```

Comment:
Sorry, I disagree. I think it's good that applications that work now will keep working with this change. The specification I wrote guarantees this: applications must currently pass input in multiples of 16 bytes, and I give them the guarantee that they don't need to start buffering output. Giving implementations extra freedom would make some applications more complex and would not have any realistic advantage for implementation design.

Reply:
I agree that applications that work now should keep working, I just think it would be clearer to document that directly, like you do for `output_size` below: (1) if only full blocks are passed, then no output is withheld, and (2) otherwise at most 15 bytes can be withheld. I think that achieves the guarantees we want and is just easier to read.

Reply:
So if I understand correctly: with inputs of length {10, 16, finish}, you want to allow outputs of length {10, 1, 15}? I don't see how this can be useful? And what about inputs of length {16, finish}: would the implementation be allowed to output {1, 15}?

Reply:
I'm afraid we're talking past each other. I don't want to allow more things. I'd just like to find the simplest wording for the documentation that still constrains the implementation enough so that applications that work now keep working.

Reply:
I'm afraid I'm more confused now. Do you want to relax the guarantees, or do you want to make the description of the guarantees easier to read? (I'm not even sure whether you two have the same answer to this.) Should we hold a meeting to discuss this (on Monday given the coming holiday week-end)?

Reply:
I'd like to make the description easier to read, including by relaxing the documented guarantees if it helps. (While making sure applications that worked are still guaranteed to work.)

But also, I don't want to delay that PR for that. I think I (we?) were trying to suggest what looked like a simple improvement, but if it take too much time to discuss, I really don't have a serious objection to the current wording.

Reply:
When I wrote my comment I had the M-AEAD PSA API in mind where there is no such details about the behavior of the implementation. Looking at it a bit more closely, the M-AEAD PSA API does not guarantee any specific number of output bytes each time an `update` is called. Output buffer of sizes PSA_AEAD_UPDATE_OUTPUT_SIZE( key_type, alg, input_length ) are of sufficient size which for GCM translates to buffer of sizes greater or equal to the multiple of 16 that is greater than `input_length` and the nearest to it thus ((input_length + 15) / 16) * 16. For the `finish` operation the output buffer size has to be greater or equal to 16.

I understand that the M-AEAD PSA API is designed for any AEAD thus that for GCM in particular the interface can be made more 'friendly' for the user. With the M-AEAD PSA API the user doesn't have much guarantee and control over `ouput_length` and with successive calls to `update` can only expect a series of output data with variable lengths.

I understand from the discussion that we would prefer to guarantee that as long as the user input buffers have a size equal to a multiple of 16 the output data sizes are exactly equal to the input buffers sizes. I think this could be guaranteed without having to define exactly the number of bytes output by `update`. Wouldn't this be enough friendly for the user?

With the current proposal I can see that even if the user has at some point input buffers with a size that is not a multiple of 16, he can manage to go back to a sequence of output data whose sizes are multiples of 16.  Do we want to enable that?

Reply:
We discussed this and decided to go with a tighter constraint, but it should be explained better. In particular, the explanation should not be split between the documentation of `output` and the documentation of `output_size`. Note that the description of `mbedtls_cipher_update` itself has a tighter constraint than just `output_length <= input_length + block_size - 1`, and we might want to use similar wording in low-level modules: “Writes as many block-sized blocks of data as possible to output.”

Reply:
I've completely reworded this part of the documentation. I decided to just spell out the two possible expected behavior, and separately document the most common strategies that callers are expected to use.